packages/amplify-provider-awscloudformation/lib/attach-backend.js

```diff
@@ -0,0 +1,299 @@
+      context.print.info(e);
+      throw e;
+    }
+  }
+
+  // If envName is not in the inputParams, prompt user to select
+  let backendEnvs = [];
+  let listEnvResponse = {};
+  do {
+    listEnvResponse = await amplifyClient
+      .listBackendEnvironments({
+        appId: amplifyApp.appId,
+        nextToken: listEnvResponse.nextToken,
+      })
+      .promise();
+
+    backendEnvs = backendEnvs.concat(listEnvResponse.backendEnvironments);
+  } while (listEnvResponse.nextToken);
+
+  if (backendEnvs.length > 1) {
+    const options = [];
+    backendEnvs.forEach(env => {
+      const option = {
+        name: env.environmentName,
+        value: env,
+        short: env.environmentName,
+      };
+      options.push(option);
+    });
+
+    const { selection } = await inquirer.prompt({
+      type: 'list',
+      name: 'selection',
+      message: 'Pick a backend environment:',
+      choices: options,
+    });
+
+    return selection;
+  } else if (backendEnvs.length === 1) {
+    context.print.info(`Found one backend environment '${backendEnvs[0].environmentName}'`);
+    const confirmMigrateMessage = 'Do you want to choose it?';
+    if (await context.prompt.confirm(confirmMigrateMessage)) {
+      return backendEnvs[0];
+    }
+  } else {
+    const errorMessage = `Found no currently existing backend environment in the amplify project: ${amplifyApp.name}.`;
+    context.print.error(errorMessage);
+    throw new Error(errorMessage);
+  }
+}
+
+async function downloadBackend(context, backendEnv, awsConfig) {
+  if (!backendEnv) {
+    return;
+  }
+  const amplifyDirPath = context.amplify.pathManager.getAmplifyDirPath(process.cwd());
+  const tempDirPath = path.join(amplifyDirPath, 'temp');
+  const currentCloudBackendDir = context.amplify.pathManager.getCurrentCloudBackendDirPath(process.cwd());
+  const backendDir = context.amplify.pathManager.getBackendDirPath(process.cwd());
+  const zipFileName = constants.S3BackendZipFileName;
+
+  const s3Client = getConfiguredS3Client(awsConfig);
+  const deploymentBucketName = backendEnv.deploymentArtifacts;
+
+  const params = {
+    Key: zipFileName,
+    Bucket: deploymentBucketName,
+  };
+
+  const zipObject = await s3Client.getObject(params).promise();
+  const buff = Buffer.from(zipObject.Body);
+
+  fs.ensureDirSync(tempDirPath);
+  const tempFilePath = path.join(tempDirPath, zipFileName);
+  fs.writeFileSync(tempFilePath, buff);
+
+  const unzippedDirPath = path.join(tempDirPath, path.basename(zipFileName, '.zip'));
+
+  await new Promise((res, rej) => {
+    extract(tempFilePath, { dir: unzippedDirPath }, err => {
+      if (err) {
+        rej(err);
+      }
+      res(unzippedDirPath);
+    });
+  });
+
+  fs.copySync(unzippedDirPath, currentCloudBackendDir);
+  fs.copySync(unzippedDirPath, backendDir);
+  fs.removeSync(tempDirPath);
+}
+
+async function getAwsConfig(context) {
```

Comment:
We already have this function, can't we reuse it?

Reply:
yes.
Moved the method to the configuration manager, and reuse it in both init and attach-backend.